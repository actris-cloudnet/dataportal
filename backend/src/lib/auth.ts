import { Connection, Repository } from 'typeorm'
import { RequestHandler } from 'express'
import { UserAccount } from '../entity/UserAccount'
const auth = require('basic-auth')
const md5 = require('apache-md5')
const { timingSafeEqual } = require('crypto')
const { Buffer } = require('buffer')

export class Authenticator {
  private userAccountRepository: Repository<UserAccount>

  constructor(conn: Connection) {
    this.userAccountRepository = conn.getRepository<UserAccount>('user_account')
  }

  middleware: RequestHandler = async (req, res, next) => {
    // Check that request has credentials included
    let credentials = auth(req)
    if (!credentials) {
      res.status(401).send('Unauthorized')
      return
    }
    let userAccount: UserAccount | undefined = await this.userAccountRepository
      .createQueryBuilder('user_account')
      .where('user_account.username = :username', { username: credentials.name })
      .getOne()
    // Check that user exists in the database
    if (userAccount === undefined || credentials.name != userAccount!.username) {
      res.status(401).send('Unauthorized')
      return
    }
    // Check that password in the request is correct
    if (
      timingSafeEqual(
        Buffer.from(md5(credentials.pass, userAccount!.passwordHash)),
        Buffer.from(userAccount!.passwordHash)
      )
    ) {
      next()
      return
    } else {
      res.status(401).send('Unauthorized')
      return
    }
  }
}
