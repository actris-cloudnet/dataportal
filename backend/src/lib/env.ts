import { URL } from "url";

const requiredVars = [
  "NODE_ENV",
  "DP_SS_URL",
  "DP_SS_USER",
  "DP_SS_PASSWORD",
  "DP_BACKEND_URL",
  "DP_FRONTEND_URL",
  "GEOLITE2_COUNTRY_PATH",
  "CITATION_SERVICE_URL",
  "DATACITE_API_URL",
  "DATACITE_API_USERNAME",
  "DATACITE_API_PASSWORD",
  "DATACITE_API_TIMEOUT_MS",
  "DATACITE_DOI_SERVER",
  "DATACITE_DOI_PREFIX",
] as const;

type RequiredVar = typeof requiredVars[number];

const missingVars = requiredVars.filter((env) => !process.env[env]);
if (missingVars.length > 0) {
  throw new Error(`FATAL: Following environment variables must be set: ${missingVars.join(", ")}`);
}

const rawEnv = process.env as NodeJS.ProcessEnv & Record<RequiredVar, string>;
const ssUrl = new URL(rawEnv.DP_SS_URL);

interface Env {
  NODE_ENV: string;
  DP_SS_URL: string;
  DP_SS_HOST: string;
  DP_SS_PORT: number;
  DP_SS_USER: string;
  DP_SS_PASSWORD: string;
  DP_BACKEND_URL: string;
  DP_FRONTEND_URL: string;
  GEOLITE2_COUNTRY_PATH: string;
  CITATION_SERVICE_URL: string;
  DATACITE_API_URL: string;
  DATACITE_API_USERNAME: string;
  DATACITE_API_PASSWORD: string;
  DATACITE_API_TIMEOUT_MS: number;
  DATACITE_DOI_SERVER: string;
  DATACITE_DOI_PREFIX: string;
}

const env: Env = {
  ...rawEnv,
  DP_SS_HOST: ssUrl.hostname,
  DP_SS_PORT: parseInt(ssUrl.port),
  DATACITE_API_URL: rawEnv.DATACITE_API_URL.replace(/\/$/, ""),
  DATACITE_API_TIMEOUT_MS: parseInt(rawEnv.DATACITE_API_TIMEOUT_MS),
  DATACITE_DOI_SERVER: rawEnv.DATACITE_DOI_SERVER.replace(/\/$/, ""),
};

export default env;
