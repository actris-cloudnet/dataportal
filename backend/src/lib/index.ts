export const stringify = (obj: any): string => JSON.stringify(obj, null, 2)
