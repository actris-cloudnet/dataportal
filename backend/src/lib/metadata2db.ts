import {createReadStream, promises as fsp} from 'fs'
import {basename} from 'path'
import {createHash} from 'crypto'
import 'reflect-metadata'
import {Connection} from 'typeorm'
import {File} from '../entity/File'
import {Site} from '../entity/Site'
import {getMissingFields, isNetCDFObject, NetCDFObject} from '../entity/NetCDFObject'
import {spawn} from 'child_process'
import {linkFile, stringify} from './index'
import config from '../config'
import {Product} from '../entity/Product'


interface NetCDFXML {
    netcdf: {
        attribute: Array<{
            '$': {
                name: string
                value: string
            }
        }>
        '$': { location: string }
    }
}


export class ReceivedFile {

  readonly ncObj: NetCDFObject
  readonly filepath: string
  readonly conn: Connection
  readonly freeze: boolean

  constructor(json: any, conn: Connection, freeze: boolean) {
    const { netcdf }: NetCDFXML = json
    const ncObj: any = netcdf.attribute
      .map((a) => a['$'])
      .map(({ name, value }) => ({ [name]: value }))
      .reduce((acc, cur) => Object.assign(acc, cur))

    if (!isNetCDFObject(ncObj)) {
      const missingFields = getMissingFields(ncObj)
      throw [`Invalid header fields\n
            Missing or invalid: ${stringify(missingFields)}\n
            ${stringify(ncObj)}`]
    }

    if (freeze && !ncObj.pid) throw ['Tried to freeze a file with no PID']

    this.ncObj = ncObj
    this.filepath = netcdf['$'].location
    this.conn = conn
    this.freeze = freeze
  }

  getUuid() {
    return this.ncObj.file_uuid
  }

  async insertFile() {
    return Promise.all([
      basename(this.filepath),
      this.computeFileChecksum(),
      this.computeFileSize(),
      this.getFileFormat(),
      this.checkSiteExists(),
      this.checkProductExists(),
      linkFile(this.filepath, config.publicDir)
    ]).then(([baseFilename, chksum, { size }, format, site, product]) => {
      const file = new File(this.ncObj, baseFilename, chksum, size, format, site, product, !this.freeze)
      file.releasedAt = new Date()
      const repo = this.conn.getRepository(File)
      return repo.save(file)
    })
  }

  async updateFile() {
    return Promise.all([
      this.computeFileChecksum(),
      this.computeFileSize(),
      this.getFileFormat()
    ]).then(([checksum, { size }, format]) => {
      const repo = this.conn.getRepository(File)
      return repo.update({uuid: this.getUuid()},
        { checksum, size, format, releasedAt: new Date(), volatile: !this.freeze, pid: (this.ncObj.pid || '') })
    })
  }

  private checkSiteExists = (): Promise<Site> =>
    this.conn.getRepository(Site).findOneOrFail(this.ncObj.location.toLowerCase().replace(/\W/g, ''))

  private checkProductExists = (): Promise<Product> =>
    this.conn.getRepository(Product).findOneOrFail(this.ncObj.cloudnet_file_type)

  private async computeFileChecksum(): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const hash = createHash('sha256')
        const input = createReadStream(this.filepath)
        input.on('readable', () => {
          const data = input.read()
          if (data)
            hash.update(data)
          else {
            const chksum = hash.digest('hex')
            resolve(chksum)
          }
        })
      } catch (err) {
        reject(err)
      }
    })
  }

  private computeFileSize() {
    return fsp.stat(this.filepath)
  }

  private getFileFormat(): Promise<string> {
    return new Promise((resolve, reject) => {
      const proc = spawn('file', [this.filepath])

      let out: string
      proc.stdout.on('data', data => out += data)
      proc.on('close', () => {
        if (out.includes('NetCDF Data Format data')) {
          resolve('NetCDF3')
        } else if (out.includes('Hierarchical Data Format (version 5) data')) {
          resolve('HDF5 (NetCDF4)')
        } else {
          reject(`Unknown file type ${out}`)
        }
      })
    })
  }
}
