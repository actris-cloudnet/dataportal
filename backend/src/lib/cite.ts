import { DataSource } from "typeorm";

import { RegularFile, ModelFile } from "../entity/File";
import { Collection } from "../entity/Collection";
import axios from "axios";
import { formatList, getCollectionLandingPage, getFileLandingPage, truncateList } from ".";
import env from "../lib/env";

const MODEL_AUTHOR: Person = { firstName: "Ewan", lastName: "O'Connor", orcid: "0000-0001-9834-5100", role: "modelPi" };
const UNKNOWN_AUTHOR: Person = { firstName: "", lastName: "CLU", orcid: null, role: "instrumentPi" };
const PUBLISHER = "ACTRIS Cloud remote sensing data centre unit (CLU)";
const MONTHS_FULL = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
];
const MONTHS_ABBR = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
const COMMON_ACK = `We acknowledge ACTRIS and Finnish Meteorological Institute for providing the data set which is available for download from <a href="${env.DP_FRONTEND_URL}">${env.DP_FRONTEND_URL}</a>.`;

export interface Person {
  firstName: string;
  lastName: string;
  orcid: string | null;
  role: "instrumentPi" | "modelPi" | "nfPi";
}

export interface Location {
  name: string;
  latitude: number | null;
  longitude: number | null;
}

export interface Citation {
  authors: Person[];
  publisher: string;
  title: string;
  year: number;
  url: string;
  note?: string;
  startDate: string;
  endDate: string;
  createdAt: string;
  updatedAt: string;
  locations: Location[];
}

interface InstrumentPi {
  first_name: string;
  last_name: string;
  orcid_id: string | null;
  start_date: string | null;
  end_date: string | null;
}

interface NfContact {
  first_name: string;
  last_name: string;
  orcid_id: string | null;
  role: string;
  start_date: string | null;
  end_date: string | null;
}

export class CitationService {
  private dataSource: DataSource;

  constructor(dataSource: DataSource) {
    this.dataSource = dataSource;
  }

  async getCollectionCitation(collection: Collection): Promise<Citation> {
    const [instrumentPis, usesModelData, nfPis, productNames, sites, dateRange] = await Promise.all([
      this.queryInstrumentPids(collection).then((pids) => fetchInstrumentPis(pids)),
      this.usesModelData(collection),
      this.queryCollectionActrisIds(collection).then((ids) => fetchNfPis(ids)),
      this.queryCollectionProductNames(collection),
      this.queryCollectionSites(collection),
      this.queryCollectionDateRange(collection),
    ]);
    if (usesModelData || productNames.includes("Model")) {
      instrumentPis.push(MODEL_AUTHOR);
    }
    const siteAuthors = await this.queryCollectionSitePersons(collection);
    const authors = removeDuplicateNames([...instrumentPis, ...nfPis, ...siteAuthors]);
    const allProducts = productNames.map(formatProductName);
    const truncatedProducts = truncateList(allProducts, 5, "products");
    let products = formatList(truncatedProducts, ", and ");
    if (allProducts.length === truncatedProducts.length) {
      products += " data";
    }
    const siteNames = sites.map((site) => site.name);
    const siteList = formatList(truncateList(siteNames, 5, "sites"), ", and ");
    const date = formatDateRange(dateRange.startDate, dateRange.endDate);
    const title = `Custom collection of ${products} from ${siteList} ${date}`;
    return {
      authors,
      publisher: PUBLISHER,
      title,
      year: collection.createdAt.getFullYear(),
      url: collection.pid || getCollectionLandingPage(collection),
      startDate: dateToString(dateRange.startDate),
      endDate: dateToString(dateRange.endDate),
      createdAt: dateToString(collection.createdAt),
      updatedAt: dateToString(collection.updatedAt),
      locations: sites,
    };
  }

  async getFileCitation(file: RegularFile | ModelFile): Promise<Citation> {
    let people: Person[];
    if (file instanceof RegularFile) {
      const [instrumentPis, nfPi, usesModelData] = await Promise.all([
        this.queryInstrumentPids(file).then((pids) => fetchInstrumentPis(pids)),
        file.site.actrisId
          ? fetchNfPi(file.site.actrisId, file.measurementDate as unknown as string)
          : Promise.resolve([]),
        this.usesModelData(file),
      ]);
      people = [
        ...instrumentPis,
        ...nfPi.map(
          (pi): Person => ({
            firstName: pi.first_name,
            lastName: pi.last_name,
            orcid: pi.orcid_id ? normalizeOrcid(pi.orcid_id) : null,
            role: "nfPi",
          }),
        ),
      ];
      if (usesModelData) {
        people.push(MODEL_AUTHOR);
      }
      people = people.concat(
        file.site.citations
          .flatMap((citation) => citation.persons)
          .map((person) => ({
            firstName: person.firstname,
            lastName: person.surname,
            orcid: person.orcid ? normalizeOrcid(person.orcid) : null,
            role: "nfPi",
          })),
      );
    } else {
      people = [MODEL_AUTHOR];
    }
    return {
      authors: removeDuplicateNames(people),
      publisher: PUBLISHER,
      title: `${file.product.humanReadableName} data from ${file.site.humanReadableName} on ${humanReadableDate(
        file.measurementDate,
      )}`,
      year: file.updatedAt.getFullYear(),
      url: file.pid || getFileLandingPage(file),
      note: file.volatile ? "Data is volatile and may be updated in the future" : undefined,
      startDate: dateToString(file.measurementDate),
      endDate: dateToString(file.measurementDate),
      createdAt: dateToString(file.createdAt),
      updatedAt: dateToString(file.updatedAt),
      locations: [],
    };
  }

  async getAcknowledgements(object: RegularFile | ModelFile | Collection): Promise<string> {
    const [siteAcks, modelAcks] = await Promise.all([this.querySiteAcks(object), this.queryModelAcks(object)]);
    let output = COMMON_ACK;
    if (siteAcks.length > 0) {
      output += " ";
      output += siteAcks.join(" ");
    }
    if (modelAcks.length > 0) {
      output += " We acknowledge ";
      output += formatList(
        modelAcks.map((a) => a.replace(/\.$/, "")),
        ", and ",
      );
      output += ".";
    }
    return output;
  }

  private queryInstrumentPids(object: RegularFile | Collection): Promise<{ instrumentPid: string; dates: string[] }[]> {
    return this.querySourceFiles(
      object,
      `SELECT instrument_info.pid AS "instrumentPid", array_agg(regular_file."measurementDate"::text) AS dates
       FROM traverse
       JOIN regular_file ON traverse.uuid = regular_file.uuid
       JOIN instrument_info ON regular_file."instrumentInfoUuid" = instrument_info.uuid
       WHERE instrument_info.pid IS NOT NULL
       GROUP BY instrument_info.pid`,
    );
  }

  private async usesModelData(object: RegularFile | Collection): Promise<boolean> {
    const rows = await this.querySourceFiles(
      object,
      `SELECT 1
       FROM traverse
       JOIN regular_file_source_model_files_model_file source_model_file ON traverse.uuid = source_model_file."regularFileUuid"
       LIMIT 1`,
    );
    return rows.length > 0;
  }

  private async queryCollectionActrisIds(collection: Collection): Promise<{ actrisId: number; dates: string[] }[]> {
    return await this.dataSource.query(
      `SELECT "actrisId", array_agg(regular_file."measurementDate"::text) AS dates
       FROM regular_file
       JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
       JOIN site ON regular_file."siteId" = site.id
       WHERE "collectionUuid" = $1
       AND "actrisId" IS NOT NULL
       GROUP BY "actrisId"`,
      [collection.uuid],
    );
  }

  private async queryCollectionProductNames(collection: Collection): Promise<string[]> {
    const rows: any[] = await this.dataSource.query(
      `SELECT product."humanReadableName"
       FROM regular_file
       JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
       JOIN product ON regular_file."productId" = product.id
       WHERE "collectionUuid" = $1
       UNION
       SELECT 'Model' AS "humanReadableName"
       FROM model_file
       JOIN collection_model_files_model_file ON model_file.uuid = "modelFileUuid"
       WHERE "collectionUuid" = $1`,
      [collection.uuid],
    );
    return rows.map((row) => row.humanReadableName).sort();
  }

  private async queryCollectionSites(collection: Collection): Promise<Location[]> {
    const rows: any[] = await this.dataSource.query(
      `SELECT site."humanReadableName", site."latitude", site."longitude"
       FROM regular_file
       JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
       JOIN site ON regular_file."siteId" = site.id
       WHERE "collectionUuid" = $1
       UNION
       SELECT site."humanReadableName", site."latitude", site."longitude"
       FROM model_file
       JOIN collection_model_files_model_file ON model_file.uuid = "modelFileUuid"
       JOIN site ON model_file."siteId" = site.id
       WHERE "collectionUuid" = $1`,
      [collection.uuid],
    );
    return rows
      .map((row) => ({
        name: row.humanReadableName,
        latitude: row.latitude,
        longitude: row.longitude,
      }))
      .sort((a, b) => a.name.localeCompare(b.name));
  }

  private async queryCollectionSitePersons(collection: Collection): Promise<Person[]> {
    const rows: any[] = await this.dataSource.query(
      `SELECT DISTINCT firstname, surname, orcid
       FROM regular_file
       JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
       JOIN site_citations_regular_citation USING ("siteId")
       JOIN regular_citation_persons_person USING ("regularCitationId")
       JOIN person ON "personId" = person.id
       WHERE "collectionUuid" = $1`,
      [collection.uuid],
    );
    return rows.map(
      (row): Person => ({
        firstName: row.firstname,
        lastName: row.surname,
        orcid: row.orcid ? normalizeOrcid(row.orcid) : null,
        role: "nfPi",
      }),
    );
  }

  private async queryCollectionDateRange(collection: Collection): Promise<{ startDate: Date; endDate: Date }> {
    const rows = await this.dataSource.query(
      `SELECT MIN("measurementDate") AS "startDate", MAX("measurementDate") AS "endDate"
       FROM (SELECT "measurementDate"
             FROM regular_file
             JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
             WHERE "collectionUuid" = $1
             UNION
             SELECT "measurementDate"
             FROM model_file
             JOIN collection_model_files_model_file ON model_file.uuid = "modelFileUuid"
             WHERE "collectionUuid" = $1) AS dates`,
      [collection.uuid],
    );
    return rows[0];
  }

  private async queryModelAcks(object: RegularFile | ModelFile | Collection): Promise<string[]> {
    if (object instanceof ModelFile) {
      return object.model.citations.map((r) => r.acknowledgements);
    }
    const rows: any[] = await this.querySourceFiles(
      object,
      `SELECT DISTINCT acknowledgements
       FROM traverse
       JOIN regular_file_source_model_files_model_file source_model_file ON traverse.uuid = source_model_file."regularFileUuid"
       JOIN model_file ON source_model_file."modelFileUuid" = model_file.uuid
       JOIN model_citations_model_citation ON model_file."modelId" = model_citations_model_citation."modelId"
       JOIN model_citation ON model_citations_model_citation."modelCitationId" = model_citation.id`,
    );
    return rows.map((row) => row.acknowledgements);
  }

  private async querySiteAcks(object: RegularFile | ModelFile | Collection): Promise<string[]> {
    if (object instanceof RegularFile) {
      return object.site.citations.map((r) => r.acknowledgements);
    }
    if (object instanceof ModelFile) {
      return [];
    }
    const rows: any[] = await this.dataSource.query(
      `SELECT DISTINCT acknowledgements
       FROM regular_file
       JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
       JOIN site_citations_regular_citation ON regular_file."siteId" = site_citations_regular_citation."siteId"
       JOIN regular_citation ON site_citations_regular_citation."regularCitationId" = regular_citation.id
       WHERE "collectionUuid" = $1`,
      [object.uuid],
    );
    return rows.map((row) => row.acknowledgements);
  }

  private async querySourceFiles(object: RegularFile | Collection, query: string): Promise<any> {
    const filter =
      object instanceof Collection
        ? `JOIN collection_regular_files_regular_file ON regular_file.uuid = "regularFileUuid"
           WHERE "collectionUuid" = $1`
        : `WHERE uuid = $1`;
    return await this.dataSource.query(
      `WITH RECURSIVE traverse AS (
         SELECT uuid
         FROM regular_file
         ${filter}
         UNION ALL
         SELECT "regularFileUuid_2"
         FROM regular_file_source_regular_files_regular_file
         JOIN traverse ON "regularFileUuid_1" = traverse.uuid
       )
       ${query}`,
      [object.uuid],
    );
  }
}

async function fetchInstrumentPi(pid: string, measurementDate?: Date): Promise<InstrumentPi[]> {
  const match = pid.match("^https?://hdl\\.handle\\.net/(.+)");
  if (!match) {
    throw new Error("Invalid PID format");
  }
  const url = `${env.HANDLE_API_URL}/handles/${match[1]}`;
  const response = await axios.get(url);
  const values = response.data.values;
  if (!Array.isArray(values)) {
    throw new Error("Invalid PID response");
  }
  const nameItem = values.find((ele) => ele.type === "URL");
  let apiUrl = nameItem.data.value + "/pi";
  if (measurementDate) {
    apiUrl += "?date=" + dateToString(measurementDate);
  }
  const apiRes = await axios.get(apiUrl);
  return apiRes.data;
}

async function fetchInstrumentPis(data: { instrumentPid: string; dates: string[] }[]): Promise<Person[]> {
  return (
    await Promise.all(
      data.map((item) =>
        fetchInstrumentPi(item.instrumentPid).then((pis) => {
          const output = new Set<InstrumentPi>();
          for (const date of item.dates) {
            pis
              .filter(
                (pi) =>
                  (pi.start_date != null ? date >= pi.start_date : true) &&
                  (pi.end_date != null ? date <= pi.end_date : true),
              )
              .forEach((pi) => {
                output.add(pi);
              });
          }
          return Array.from(
            output,
            (pi): Person => ({
              firstName: pi.first_name,
              lastName: pi.last_name,
              orcid: pi.orcid_id ? normalizeOrcid(pi.orcid_id) : null,
              role: "instrumentPi",
            }),
          );
        }),
      ),
    )
  ).flat();
}

async function fetchNfPi(actrisId: number, measurementDate?: string): Promise<NfContact[]> {
  const response = await axios.get(`${env.LABELLING_URL}/api/facilities/${actrisId}/contacts`, {
    params: {
      date: measurementDate,
      role: "pi",
    },
  });
  if (!Array.isArray(response.data)) {
    return [];
  }
  return response.data;
}

async function fetchNfPis(data: { actrisId: number; dates: string[] }[]): Promise<Person[]> {
  return (
    await Promise.all(
      data.map((item) =>
        fetchNfPi(item.actrisId).then((contacts) => {
          const output = new Set<InstrumentPi>();
          for (const date of item.dates) {
            contacts
              .filter(
                (contact) =>
                  contact.role === "pi" &&
                  (contact.start_date != null ? date >= contact.start_date : true) &&
                  (contact.end_date != null ? date <= contact.end_date : true),
              )
              .forEach((pi) => {
                output.add(pi);
              });
          }
          return Array.from(
            output,
            (pi): Person => ({
              firstName: pi.first_name,
              lastName: pi.last_name,
              orcid: pi.orcid_id ? normalizeOrcid(pi.orcid_id) : null,
              role: "nfPi",
            }),
          );
        }),
      ),
    )
  ).flat();
}

function normalizeOrcid(orcid: string): string {
  return orcid.replace(/^(https?:\/\/)?(www\.)?orcid\.org\//, "");
}

function normalizeText(input: string): string {
  return input
    .trim()
    .normalize("NFKD")
    .replace(/[\u0300-\u036F]/g, "")
    .toLowerCase();
}

export function removeDuplicateNames(pis: Person[]): Person[] {
  // Order
  const allPis = pis
    .sort((a, b) =>
      a.lastName !== b.lastName
        ? a.lastName.localeCompare(b.lastName, "en-gb")
        : a.firstName.localeCompare(b.firstName, "en-gb"),
    )
    .filter((ele) => ele.role == "instrumentPi")
    .concat(pis.filter((ele) => ele.role == "modelPi"))
    .concat(pis.filter((ele) => ele.role == "nfPi"));
  // Remove duplicates
  const out: Person[] = [];
  for (const pi of allPis) {
    const nameExists = pi.orcid
      ? out.some((name) => name.orcid === pi.orcid)
      : out.some(
          (name) =>
            normalizeText(name.firstName) === normalizeText(pi.firstName) &&
            normalizeText(name.lastName) === normalizeText(pi.lastName),
        );
    if (!nameExists) {
      out.push(pi);
    }
  }
  // DataCite requires at least one author.
  if (out.length === 0) {
    out.push(UNKNOWN_AUTHOR);
  }
  return out;
}

function formatDateRange(startDate: Date, endDate: Date): string {
  const sameDate = startDate.getDate() == endDate.getDate();
  const sameMonth = startDate.getMonth() == endDate.getMonth();
  const sameYear = startDate.getFullYear() == endDate.getFullYear();
  if (sameYear && sameMonth && sameDate) {
    return "on " + startDate.getDate() + " " + MONTHS_ABBR[startDate.getMonth()] + " " + startDate.getFullYear();
  }
  let output = "between " + startDate.getDate();
  if (!sameMonth || !sameYear) {
    output += " " + MONTHS_ABBR[startDate.getMonth()];
  }
  if (!sameYear) {
    output += " " + startDate.getFullYear();
  }
  return output + " and " + endDate.getDate() + " " + MONTHS_ABBR[endDate.getMonth()] + " " + endDate.getFullYear();
}

function humanReadableDate(date: Date): string {
  if (!(date instanceof Date)) {
    date = new Date(date);
  }
  return `${date.getDate()} ${MONTHS_FULL[date.getMonth()]} ${date.getFullYear()}`;
}

const capitalWords = ["Doppler", "MWR", "(Voodoo)", "L3"];

function formatProductName(product: string) {
  return product
    .split(" ")
    .map((word) => (capitalWords.includes(word) ? word : word.toLowerCase()))
    .join(" ");
}

function dateToString(date: Date | string) {
  return new Date(date).toISOString().slice(0, 10);
}
