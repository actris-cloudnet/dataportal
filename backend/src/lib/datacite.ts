import { Collection } from "../entity/Collection";
import axios from "axios";
import env from "./env";
import { CitationService } from "./cite";
import { getCollectionLandingPage } from ".";

export class DataCiteService {
  private citationService: CitationService;

  constructor(citationService: CitationService) {
    this.citationService = citationService;
  }

  async createCollectionDoi(collection: Collection): Promise<string> {
    if (collection.pid) {
      throw new Error(`Collection ${collection.uuid} already has a PID: ${collection.pid}`);
    }
    const res = await axios.post(`${env.DATACITE_API_URL}/dois`, await this.collectionDataCite(collection), {
      headers: { "Content-Type": "application/vnd.api+json" },
      auth: { username: env.DATACITE_API_USERNAME, password: env.DATACITE_API_PASSWORD },
      timeout: env.DATACITE_API_TIMEOUT_MS,
    });
    return `${env.DATACITE_DOI_SERVER}/${res.data.data.attributes.doi}`;
  }

  async updateCollectionDoi(collection: Collection) {
    if (!collection.pid || !collection.pid.startsWith(`${env.DATACITE_DOI_SERVER}/${env.DATACITE_DOI_PREFIX}/`)) {
      throw new Error(`Collection ${collection.uuid} doesn't have a valid PID: ${collection.pid}`);
    }
    const doi = collection.pid.slice(env.DATACITE_DOI_SERVER.length + 1);
    await axios.put(`${env.DATACITE_API_URL}/dois/${doi}`, await this.collectionDataCite(collection), {
      headers: { "Content-Type": "application/vnd.api+json" },
      auth: { username: env.DATACITE_API_USERNAME, password: env.DATACITE_API_PASSWORD },
      timeout: env.DATACITE_API_TIMEOUT_MS,
    });
  }

  private async collectionDataCite(collection: Collection): Promise<object> {
    const doiSuffix = collection.uuid.toLowerCase().replace(/-/g, "").slice(0, 16);
    const citation = await this.citationService.getCollectionCitation(collection);
    const creators = citation.authors.map((person) => ({
      name: `${person.lastName}, ${person.firstName}`,
      nameType: "Personal",
      givenName: person.firstName,
      familyName: person.lastName,
      nameIdentifiers: person.orcid
        ? [
            {
              schemeUri: "https://orcid.org",
              nameIdentifier: `https://orcid.org/${person.orcid}`,
              nameIdentifierScheme: "ORCID",
            },
          ]
        : undefined,
    }));
    return {
      data: {
        type: "dois",
        attributes: {
          event: "publish",
          doi: `${env.DATACITE_DOI_PREFIX}/${doiSuffix}`,
          creators,
          titles: [{ lang: "en", title: citation.title }],
          publisher: citation.publisher,
          publicationYear: citation.year,
          types: { resourceTypeGeneral: "Dataset" },
          url: getCollectionLandingPage(collection),
          schemaVersion: "http://datacite.org/schema/kernel-4",
          language: "en",
          dates: [
            { date: citation.createdAt, dateType: "Created" },
            { date: `${citation.startDate}/${citation.endDate}`, dateType: "Collected" },
          ],
          formats: ["application/zip", "application/netcdf"],
          rightsList: [
            {
              lang: "en",
              schemeURI: "https://spdx.org/licenses/",
              rightsIdentifierScheme: "SPDX",
              rightsIdentifier: "CC-BY-4.0",
              rightsURI: "https://creativecommons.org/licenses/by/4.0/",
              rights: "Creative Commons Attribution 4.0 International",
            },
          ],
          geoLocations: citation.locations.map((location) => ({
            geoLocationPlace: location.name,
            geoLocationPoint:
              location.latitude && location.longitude
                ? {
                    pointLatitude: location.latitude,
                    pointLongitude: location.longitude,
                  }
                : undefined,
          })),
        },
      },
    };
  }
}
