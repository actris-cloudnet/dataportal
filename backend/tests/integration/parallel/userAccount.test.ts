import axios from "axios";
import { backendPublicUrl } from "../../lib";
import { describe, expect, it } from "@jest/globals";

const url = `${backendPublicUrl}users/me/`;
const headers = { "content-type": "application/json" };

describe("GET /users/me", () => {
  it("returns unauthorized error", async () => {
    await expect(axios.get(url, { headers, auth: { username: "hax0r", password: "hax0r" } })).rejects.toMatchObject({
      response: { status: 401 },
    });
  });

  it("return submitter permissions", async () => {
    const res = await axios.get(url, { headers, auth: { username: "submitter", password: "submitter" } });
    expect(res.status).toEqual(200);
    expect(res.data).toMatchSnapshot();
  });

  it("returns admin permissions", async () => {
    const res = await axios.get(url, { headers, auth: { username: "admin", password: "admin" } });
    expect(res.status).toEqual(200);
    expect(res.data).toMatchSnapshot();
  });
});
