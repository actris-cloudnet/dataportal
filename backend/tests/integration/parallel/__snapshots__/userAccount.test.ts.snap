// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /users/me return submitter permissions 1`] = `
[
  {
    "id": 9,
    "permission": "canUpload",
    "site": {
      "actrisId": null,
      "altitude": 174,
      "country": "Finland",
      "countryCode": "FI",
      "countrySubdivisionCode": null,
      "description": null,
      "dvasId": null,
      "gaw": "SMR",
      "humanReadableName": "Hyytiälä",
      "id": "hyytiala",
      "latitude": 61.844,
      "longitude": 24.288,
      "type": [
        "cloudnet",
      ],
    },
  },
]
`;

exports[`GET /users/me returns admin permissions 1`] = `
[
  {
    "id": 1,
    "permission": "canUpload",
    "site": null,
  },
  {
    "id": 2,
    "permission": "canUploadModel",
    "site": null,
  },
  {
    "id": 3,
    "permission": "canDelete",
    "site": null,
  },
  {
    "id": 4,
    "permission": "canProcess",
    "site": null,
  },
  {
    "id": 5,
    "permission": "canCalibrate",
    "site": null,
  },
  {
    "id": 6,
    "permission": "canGetStats",
    "site": null,
  },
  {
    "id": 7,
    "permission": "canAddPublication",
    "site": null,
  },
  {
    "id": 8,
    "permission": "canPublishTask",
    "site": null,
  },
]
`;
